src/Microsoft.Sbom.Parsers.Spdx22SbomParser/SPDXParser.cs

```diff
@@ -159,8 +172,77 @@ ParserState MoveToNextState()
+        do
+        {
+            internalParserState = ProcessPropertyInternal();
+        }
+        while (internalParserState == ParserState.INTERNAL_METADATA);
+
+        return internalParserState;
+
+        ParserState ProcessPropertyInternal()
+        {
+            var reader = new Utf8JsonReader(buffer, isFinalBlock: isFinalBlock, readerState);
+            
+            // The root properties parser consumes the value of the property as well. For example,
+            // "spdxId": "SPDXID", root parser would have already consumed the SPDXID string. To 
+            // work around this issue, the root parser will return the next token which we store in 
+            // nextTokenString and use it here instead of doing a reader.ReadString().
+            switch (currentRootPropertyName)
+            {
+                case Constants.SPDXVersionHeaderName:
+                    metadata.SpdxVersion = nextTokenString;
+                    break;
+                case Constants.DataLicenseHeaderName:
+                    metadata.DataLicense = nextTokenString;
+                    break;
+                case Constants.DocumentNameHeaderName:
+                    metadata.Name = nextTokenString;
+                    break;
+                case Constants.DocumentNamespaceHeaderName:
+                    if (string.IsNullOrEmpty(nextTokenString))
+                    {
+                        throw new ParserException($"Document namespace URI is null or empty.");
+                    }
+
+                    metadata.DocumentNamespace = new Uri(nextTokenString);
+                    break;
+                case Constants.CreationInfoHeaderName:
+                    var parser = new CreationInfoParser(stream);
+                    metadata.CreationInfo = parser.GetCreationInfo(ref buffer, ref reader);
+                    break;
+                case Constants.DocumentDescribesHeaderName:
+                    metadata.DocumentDescribes = ParserUtils.ParseListOfStrings(stream, ref reader, ref buffer);
+                    break;
+                case Constants.SPDXIDHeaderName:
+                    metadata.SpdxId = nextTokenString;
+                    break;
+                default:
+                    throw new ParserException($"Unknown metadata property {currentRootPropertyName} found while parsing metadata.");
+            }
+
+            var rootPropertiesParser = new RootPropertiesParser(stream);
+            var result = rootPropertiesParser.MoveNext(ref buffer, ref reader);
+            ParserUtils.GetMoreBytesFromStream(stream, ref buffer, ref reader);
+
+            isFinalBlock = reader.IsFinalBlock;
+            readerState = reader.CurrentState;
+            currentRootPropertyName = result.PropertyName;
+            nextTokenString = result.NextToken;
+            return result.State;
+        }
+    }
+
     private void ValidateParsingComplete()
     {
+        if (!metadataStateProcessed)
```

Comment:
NitPick: I don't think this actually validates anything important to the parser. As far as I can tell it's just mandating that  the user calls `GetMetadata`. That makes sense for `GetReferences`, `GetPackages`, etc because they can control the state of the stream/parser, but the metadata is already parsed regardless, it's just a matter of whether the user views the result or not.

Reply:
The way the state machine is designed, the only way to get to the finished state without a failure event is to get the metadata. Since we want that to be called last. So this becomes necessary.